proposals/NNNN-isolated-default-values.md

```diff
@@ -0,0 +1,173 @@
+
+Note that the inferred actor isolation of a closure literal can never be an actor instance based on the following two properties:
+
+1. To be isolated to an actor instance, a closure must either have its own (explicit) isolated parameter or capture an isolated parameter from its enclosing context.
+2. Closure literals in default arguments cannot capture values.
+
+### Enforcing default value isolation requirements
+
+#### Default argument values
+
+Isolation requirements for default argument expressions are enforced at the caller. If the caller is not in the required isolation domain, the default argument cannot be used and the argument must be specified explicitly. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefault(value: Int = requiresMainActor()) { ... }
+
+@MainActor func mainActorCaller() {
+  useDefault() // okay
+}
+
+func nonisolatedCaller() async {
+  useDefault() // error
+
+  useDefault(value: await requiresMainActor()) // okay
+}
+```
+
+#### Stored property initial values
+
+Isolation requirements for default initializer expressions for stored properties apply in the body of initializers. If an `init` does not match the isolation of the initializer expression, the initialization of that stored property is not emitted at the beginning of the `init`. Instead, the stored property must be explicitly initialized in the body of the `init`. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+@AnotherActor func requiresAnotherActor() -> Int { ... }
+
+class C {
+  @MainActor var x1: Int = requiresMainActor()
+  @AnotherActor var x2: Int = requiresAnotherActor()
+
+  nonisolated init() {} // error
+
+  nonisolated init(x1: Int, x2: Int) { // okay
+    self.x1 = x1
+    self.x2 = x2
+  }
+
+  @MainActor init(x2: Int) { // okay
+    // 'self.x1' gets assigned to the default value 'requiresMainActor()'
+    self.x2 = x2
+  }
+}
+```
+
+In the above example, the no-parameter `nonisolated init()` is invalid, because it does not initialize `self.x1` and `self.x2`. Because the default initializer expressions require different actor isolation, those values are not used in the `nonisolated` initializer. The other two initializers are valid.
+
+### Default value isolation in memberwise initializers
+
+For structs, default initializer expressions for stored properties are used as default argument values to the compiler-generated memberwise initializer. In this case, the default argument value shares the same required isolation as the default initializer expression. Because the default values are always evaluated in the caller's context, all the memberwise initializer does is initialize each field, which can always be performed in a `nonisolated` context. In the interest of only applying global actor isolation when it's necessary for the code to run on the global actor, this proposal also changes the isolation of memberwise initializers to be `nonisolated`.
```

Comment:
This would benefit from an example that shows what the synthesized memberwise initializer looks like.

Reply:
Alos: classes have a no-argument initializer when their superclass does and all of their stored properties have default values. I suppose we need to apply the same rules there?

Reply:
My thinking is the no-argument initializer would still get the actor isolation of the class, because it's so common for default value expressions to have the same isolation. I suppose we could make the no-argument initializer `nonisolated` if all default value expressions are `nonisolated`. Either way, I should specify the behavior of the no-argument initializer in the proposal so it can be discussed on the forums.

Reply:
I've expanded the section on compiler-synthesized initializers and included a few examples. The new rules are a little nuanced though, so please let me know if this needs more clarification.